### ISSUE TITLE
Move definitions of BeforeBuild and AfterBuild targets to .props instead of .targets files

### ISSUE THREAD

**Issue by dsplaisted** [Microsoft/msbuild#1680]
The `BeforeBuild` and `AfterBuild` targets are currently defined in [Microsoft.Common.CurrentVersion.targets](https://github.com/Microsoft/msbuild/blob/ca7b47d21088068aa7f8055b5637c5eb5ab500db/src/XMakeTasks/Microsoft.Common.CurrentVersion.targets#L783-L792).  I presume that the intention was for you to override them in your project files after the .targets import near the bottom of the file.

However, if you're using the new `Sdk` attribute on the `Project` element, it's not possible to put a target definition after the default .targets import.  This can lead to targets that people put in their project files unexpectedly not running, with no indication why unless you examine the log file and see the message that the target has been overridden (for example, https://github.com/dotnet/sdk/issues/841).

It would be better to define the empty `BeforeBuild` and `AfterBuild` targets in a .props file so that if they occur in the body of a project the ones from the project take precedence.

@AndyGerlicher @rainersigwald @cdmihai What do you think about the compat implications of this and when we could make such a change?  If we changed it for all situations, then targets defined in the "wrong" place in project files would start running where they hadn't previously.  If we are not OK with that, we could change to conditionally defining these targets where they currently are, and then define them in a .props file of the .NET SDK along with a property telling the default MSBuild targets not to define them.

**Comment by cdmihai** [comment]
As a workaround for now, is to use SDK imports to get around this:

```xml
<Project>
  <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk/1.0.0" />
...
  <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk/1.0.0" />

  <Target Name="BeforeBuild">
...
  </Target>
</Project>
```

**Comment by rainersigwald** [comment]
I'm very worried about the compat impact of moving things around in common targets. I do not think we should make this change, ever.

The condition on a target is considered when the target is run, not when it is evaluated. That means there's no such thing as a conditional target definition, so there's no way to mitigate the compat impact as you describe.

Additionally, I do not think we should make changes to encourage the use of the confusing override-`BeforeBuild`/`AfterBuild` extensibility mechanism over the clearer alternative of adding a new target that runs at the right time defined by `AfterTargets` or `BeforeTargets`. I know that the comment in the old default csproj template encouraged this. That was wrong then and is wrong now.

If a user really wants to override an SDK-delivered target, they should switch to the explicit import form to make the ordering clear, as @cdmihai showed.

**Comment by dsplaisted** [comment]
I wasn't explicit in the original issue description, but you can do a conditional target definition by putting the target in a separate file and conditionally importing it.  That's what I was thinking of doing for this.

I think it's going to be common for people to hit this as they port projects to the new SDK.  If you just had to add `AfterTargets="Build"` to your target, that wouldn't be bad, but you also have to choose a different name for the target, when `AfterBuild` or `BeforeBuild` are the natural choices.  So we end up with [`MyAfterBuild` targets](https://github.com/dotnet/sdk/blob/faac7c6ada41a1ab92857dcbbb9d6feba0dbd9a4/src/Dependencies/xUnit.net/xUnit.net.csproj#L19) in the projects.

**Comment by rainersigwald** [comment]
I couldn't disagree more strongly with this:

> `AfterBuild` or `BeforeBuild` are the natural choices

The natural name for a target, like the natural name for a function, is a description of what the target does--for all the same reasons. I would hard reject any code review I saw with a pattern like `MyAfterBuild`.

**Comment by dsplaisted** [comment]
Good point that BeforeBuild and AfterBuild are not great names.  Unfortunately we've trained people to use them over 10 or so years, so in that sense I do think they are names that people will naturally use.

In general, I think we should prefer avoiding friction when adopting the new Sdk-style projects, as opposed to imposing better practices on patterns that worked fine before.

**Comment by rainersigwald** [comment]
That's definitely the line of argument that's going to convince me, if anything will. But it doesn't yet.

I don't want to saddle all new projects with the failures of the past. We've taken the new Sdk as a place to make many breaks from old behavior, and I think should be one of them.

**Comment by karolz-ms** [comment]
As Daniel said, this is going to break a lot of people and there is a lot of docs on the net telling people to use these magic names. Just run into this myself--very confusing.

If we do not want the magic, at least we should make sure that BeforeBuild/AfterBuild names are not "cursed" (they currently are, meaning, if your target is named AfterBuild, it won't be executed even if it has AfterTargets attribute set). This feels just mean and silly.

**Comment by jayoungers** [comment]
I've just started switching my project file formats over and thank god I ran into these threads: for the life of me I couldn't figure out why none of my build scripts were running.

Going from `Target Name="AfterCompile"` to `Target Name="MyTask" AfterTargets="Compile"` or from `Target Name="AfterBuild"` to `Target Name="MyTask" AfterTargets="Build"` resolved all of my issues.

Like was mentioned, from all of the tutorials out there I just assumed those special names were how you were supposed to setup these.

**Comment by davkean** [comment]
I want to add my 2cents; I've been involved in this project from day 1 - yet I was utterly confused when my `AfterBuild` target didn't run. If someone closely involved in the project runs into this - I can't imagine how 3rd parties can understand this change.

**Comment by terrajobst** [comment]
What is the recommended replacement?

```xml
<Target Name="MyCode" AfterTargets="AfterBuild" />
```

-- or --

```xml
<Target Name="MyCode" AfterTargets="Build" />
```

-- or --

```xml
<PropertyGroup>
  <BuildDependsOn>$(BuildDependsOn);MyCode</BuildDependsOn>
</PropertGroup>

<Target Name="MyCode" />
```

They aren't equivalent and have different behavior (and there are probably more ways to do that). We should agree what the right replacement is and recommend it in our docs.

**Comment by rainersigwald** [comment]
The most direct replacement is 

```xml
<Target Name="MyCode" BeforeTargets="Build" />
```

(note `Before`, not `After`).

That is often appropriate for a "just needs to get done eventually" target. In many cases, I would recommend using `AfterTargets` on the target that produces the output that needs to be modified instead, for better understandability and future extensibility. For example, you might have

```csproj
<Target Name="RewriteILUsingMagic" AfterTargets="CoreCompile">
  <Exec Command="magic.exe @(IntermediateRefAssembly) -rewrite" />
</Target>
```

**Comment by terrajobst** [comment]
So for `AfterBuild` it sounds like the recommendation is

```xml
<Target Name="MyCode" BeforeTargets="Build" />
```

What about `BeforeBuild`?

**Comment by rainersigwald** [comment]
```xml
<Target Name="MyCode" BeforeTargets="CoreBuild" />
```

Is the direct replacement, though in this case I'd recommend even more strongly that you should refer directly to the target that will consume the output you're producing (perhaps `BeforeTargets="CoreCompile"` if you're generating code).

**Comment by terrajobst** [comment]
Makes sense. I supposed it would be useful to have a table with a few entries, plus the recommendation to check `Microsoft.Common.targets` for details, as most of it is probably not going to be documented for now...

**Comment by magol** [comment]
One issue i have with AfterBuild is when i use multiple frameworks. The target is run efter each target framework has been compiled. But how do i do if i want a target to run after all target has been compiled?

**Comment by sayedihashimi** [comment]
If we don't want to move forward `BeforeBuild`/`AfterBuild` that's one thing, but users should be able to name the targets `BeforeBuild`/`AfterBuild` as they have before. Today if you create the following target in your .csproj

```
<Target Name="BeforeBuild" BeforeTargets="Build">
  <Message Text="Inside BeforeBuild" Importance="high"/>
</Target>
```

The target will not run and you'll see the following in the log file.

```
Overriding target "BeforeBuild" in project "C:\Users\sayedha\source\repos\TuneTag\TuneTag
\TuneTag.csproj" with target "BeforeBuild" from project "C:\Program Files (x86)\Microsoft Visual 
Studio\2017\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets".
```

What's the value of defining an empty target in `Microsoft.Common.CurrentVersion.targets` for `BeforeBuild` and `AfterBuild`?

Also I'd like to echo @davkean comment.

>I want to add my 2cents; I've been involved in this project from day 1 - yet I was utterly confused when my AfterBuild target didn't run. If someone closely involved in the project runs into this - I can't imagine how 3rd parties can understand this change.

I was really confused when my `BeforeBuild` target didn't execute, especially after adding `AfterTargets`. I don't think a typical customer would be able to figure out that they would need to change the name of the target. It just appears as yet another thing that doesn't work with core.

I understand we would like customers to move away from BeforeBuild/AfterBuild but there is an extensive amount of docs/info available that point users to that.

**Comment by MarkusAmshove** [comment]
I've migrated a solution with the old csproj format to the new one, but the lack of `BeforeBuild` did break my build.
I now have a mix of old and new format in that solution, because a project depends on a code generation tool to run before the compiler kicks in.
Is there any way (in the new format) to force a traget to be run before the compiler kicks in, without having to specify the target directly via MSBuild flags (because the build in VS should do it by default too)?

The workaround with import targets and props doesn't work for me anymore, just giving some warnings and not invoking the target.

**Comment by weltkante** [comment]
You did not only break BeforeBuild/AfterBuild but also BuildDependsOn. And probably all other {Target}DependsOn Variables defined in MSBuild, like CoreBuildDependsOn, RunDependsOn, and whatever else there may exist. Now we have to redo our entire Build Process which worked fine for classic projects for 10+ years. Very annoying.

@rainersigwald 
> I'm very worried about the compat impact of moving things around in common targets. I do not think we should make this change, ever.

The compact impact of *not* moving them around is that all After{Target}, Before{Target} and {Target}DependsOn Variables do not work anymore at all. Any targets files which made use of these are now subtly broken. Think of targets inside nuget packages which may be included in either classic or new project types.

**Comment by vatsan-madhavan** [comment]
We just had a project migrated that had a codegen target defined like this that failed to run: 

```
<PropertyGroup>
<CoreCompileDependsOn>
GenerateTemplates;
$(CoreCompileDependsOn);
</CoreCompileDependsOn>
</PropertyGroup>
```

I don't `CoreCompileDependsOn` called out on this thread explicitly - is this a variant of the problem being discussed here? 

/cc @danzil

**Comment by davkean** [comment]
Yes.

**Comment by vatsan-madhavan** [comment]
Do we have a plan for prevent these sorts of surprises? When projects are migrated over to Sdk style csproj, we should expect these to just work. This feels like a significant compatibility loss to me - esp. one that can often be a silent problem in sufficiently complex systems.

**Comment by rainersigwald** [comment]
As a general rule, migrated projects should use explicit SDK imports, rather than the implicit form, for exactly these reasons. That is, replace the old project template's

```xml
<Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
```

with

```xml
<Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
```

and likewise replace

```xml
<Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
```

with 

```xml
<Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
```

In that way, the import order is preserved, allowing properties and targets to be overridden as they have been historically.

**Comment by weltkante** [comment]
@rainersigwald I'm not sure its just a problem for project migration, what about targets imported from nuget packages, like code generators?
 
I worry that this can subtly break packages which rely on After{Target}, Before{Target} or {Target}DependsOn variables. It definitely did happen for our own build customization and we didn't notice for months that some rules did not run anymore when imported into SDK projects.

For us it basically means we cannot use implicit SDK projects at all (regardless of whether they are migrated or newly created).

**Comment by rainersigwald** [comment]
@weltkante Can you please provide a concrete example? The relative import order of NuGet-package-delivered imports should not change between SDK and non-SDK projects (the "import all NuGet .targets" import comes after the common.currentversion.targets import), so I don't think I understand the problem you're describing.

Note also that it's bad practice for NuGet packages to override an extension target like `AfterBuild`, independently of SDK projects--what if you want to use two such packages?

**Comment by weltkante** [comment]
> the "import all NuGet .targets" import comes after the common.currentversion.targets import

I wasn't aware of the concrete ordering, sorry, I guess that means it should work. Our scenario was a manually referencing a targets file and I was assuming that targets referenced by nuget packages may have the same problem, but I didn't try to build an example. Given the ordering you mention it makes sense that only the explicit references in the project file itself are broken and not those included by other mechanics.

> Note also that it's bad practice for NuGet packages to override an extension target like `AfterBuild`, independently of SDK projects--what if you want to use two such packages?

`{Target}DependsOn` is composable but suffers from the same bug. You include the previous value of the variable when extending it.

**Comment by rainersigwald** [comment]
> `{Target}DependsOn` is composable (you include the previous value of the variable when extending it) but suffers from the same bug.

That should be handled correctly for NuGet packages by the import order, too, and can be handled with explicit Sdk imports for the manual (or ported-from-non-SDK-project) case.

**Comment by herebebeasties** [comment]
The recommendation earlier in this thread that you use:
```
<Target Name="MyCode" BeforeTargets="Build" />
```
As a replacement for the `AfterBuild` target looks plain wrong. Not only would that read very confusingly, but it certainly does the wrong thing. The `Build` target itself does nothing. It is just composed of three other BeforeTargets that therefore do the work: `BeforeBuild`, `CoreBuild` and `AfterBuild`.

If you make your target depend have `BeforeTargets="Build"` it will obviously therefore run _before_ the `CoreBuild` target (that itself depends on targets that depend on `CoreCompile`, etc.) rather than after it.

The solution for this is therefore to rewrite your BeforeBuild / AfterBuild target overrides to simply be, respectively:
```
<Target Name="CustomBeforeBuild" BeforeTargets="Build" />
<Target Name="CustomAfterBuild" AfterTargets="Build" />
```
This will then compose properly, read well and behave as you expect.

**Comment by rainersigwald** [comment]
@herebebeasties That's not correct. The docs describe [the exact target-ordering algorithm](https://docs.microsoft.com/en-us/visualstudio/msbuild/target-build-order?view=vs-2017#determine-the-target-build-order), but the critical part is that `BeforeTargets` is considered _after_ `DependsOnTargets`.

To prove this to yourself here's a simple example:

```xml
<Project DefaultTargets="Build">
 <Target Name="Build" DependsOnTargets="BeforeBuild;CoreBuild;AfterBuild" />

 <Target Name="BeforeBuild" />
 <Target Name="CoreBuild" />
 <Target Name="AfterBuild" />

 <Target Name="Custom" BeforeTargets="Build" />
</Project>
```

```
S:\msbuild>msbuild -verbosity:detailed ordering.proj
Microsoft (R) Build Engine version 16.0.443+g5775d0d6bb for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 3/1/2019 9:46:14 AM.
     1>Project "S:\msbuild\ordering.proj" on node 1 (default targets).
     1>Building with tools version "Current".
     1>Target "BeforeBuild" in project "S:\msbuild\ordering.proj" (target "Build" depends on it):
     1>Done building target "BeforeBuild" in project "ordering.proj".
     1>Target "CoreBuild" in project "S:\msbuild\ordering.proj" (target "Build" depends on it):
     1>Done building target "CoreBuild" in project "ordering.proj".
     1>Target "AfterBuild" in project "S:\msbuild\ordering.proj" (target "Build" depends on it):
     1>Done building target "AfterBuild" in project "ordering.proj".
     1>Target "Custom" in project "S:\msbuild\ordering.proj" (target "Build" depends on it):
     1>Done building target "Custom" in project "ordering.proj".
     1>Target "Build" in project "S:\msbuild\ordering.proj" (entry point):
     1>Done building target "Build" in project "ordering.proj".
     1>Done Building Project "S:\msbuild\ordering.proj" (default targets).

Build succeeded.
```

I recommend against ever using `AfterTargets="Build"`. It has very confusing semantics and buggy behavior in multiproc builds (#3345).

If you wish to be very explicit in ordering,

```xml
<Target Name="CustomBeforeBuild" BeforeTargets="BeforeBuild" />
<Target Name="CustomAfterBuild" AfterTargets="AfterBuild" />
```

is a better approach.

**Comment by herebebeasties** [comment]
Thank you very much for clarifying this. I hadn't realised the ordering there, which obviously makes sense when you think about it, but nonetheless gives a rather counterintuitive result that a <Target> with BeforeTargets="Build" runs after the build. We'll use your suggestion as it's less likely to be inadvertently changed by other devs to "fix" it.

**Comment by nguerrera** [comment]
Going way back to https://github.com/Microsoft/msbuild/issues/1680#issuecomment-278504147

> That means there's no such thing as a conditional target definition, so there's no way to mitigate the compat impact as you describe.

Isn't this possible like so:

#### Place where you want to put target but only on some condition:
``` xml
<Import Project="ConditionalTargets.targets" Condition="$(SomeCondition)" />
```

#### ConditionalTargets.targets
``` xml
<Target Name="ConditionalTarget" />
```

**Comment by rainersigwald** [comment]
Yes, that's true.

**Comment by nguerrera** [comment]
Using that could we maybe do this:

1. Have a targets file with all the overridable stubs in common.targets. Have it set a flag that it has already been imported.
2. In common.props, set a variable to the path of this targets file.
3. In common.targets, import it conditioned on whether it's already been imported.
4. In sdk.props for 3.0, import the stubs early.

This should be compatible with sdk < 2.0 and all classic projects, and only break some exotic cases going from 2.0 to 3.0 SDK.

**Comment by rainersigwald** [comment]
My initial reaction to that plan is utter revulsion, followed by agreeing that we should do it since it's a bunch of gymnastics we can do to avoid user pain. I think that's the plan now.

**Comment by sharwell** [comment]
I would be more in favor of making these targets work if they were originally given the more accurate names `MaybeBeforeBuild` and `MaybeAfterBuild`. 😄

**Comment by NinoFloris** [comment]
Before this becomes legacy that carries over into new sdk forever. Would a new sdk.compat that does that and possibly more gymnastics not be preferable? Then there's at least some hint that at some point in the future this could cease to exist. Nudging people to gradual migration

**Comment by dashesy** [comment]
We have a common dependency nuget that have this in the nuspec

```XML
  <Target Name="CreateModelConfig" AfterTargets="Build" Outputs="%(Model.Identity)">
  <SomeTask Input="%(Model.Input)"/>
  </Target>
```

Should we change it to `AfterTargets="CoreCompile"` if we want to to be usable in VS2019?

**Comment by weltkante** [comment]
@dashesy no, `AfterTargets` works just fine, this issue is about `Before{TargetName}` and `After{TargetName}` _properties_ (which are global lists of targets) not about the literal `BeforeTargets` and `AfterTargets` _attributes_ (which are attributes on the target itself).

**Comment by rainersigwald** [comment]
@dashesy In general I recommend that you change to `BeforeTargets="AfterBuild"` (or a more specific target, depending on details).

The `Build` target is the very last thing to run in the default configuration; it does nothing and exists only to call `BeforeBuild`, `CoreBuild`, and `AfterBuild` targets (in that order). Hooking after Build has confusing semantics because it means work happens after the project is "done". That leads to bugs like #3345.

I would not recommend `AfterTargets="CoreCompile"` because it makes it harder for a user to inject itself into the middle (if for example they want to use an IL rewriter after the compiler but before packaging).

**Comment by ericsampson** [comment]
@rainersigwald is any work planned to help people migrating avoid this pain? It's so easy to get tripped up.

**Comment by aaronla-ms** [comment]
Ditto. We've solved it locally by adding the following to the root Directory.Build.targets our repos:

```
<!-- 
Alternate beforebuild/afterbuild that's compatible with both old style projects and sdk projects. See
issue https://github.com/dotnet/msbuild/issues/1680 for details: 
-->
<Target Name="XyzBeforeBuild" BeforeTargets="BeforeBuild" DependsOnTargets="$(BeforeBuildDependsOn)" />
<Target Name="XyzAfterBuild" BeforeTargets="AfterBuild" DependsOnTargets="$(AfterBuildDependsOn)" />
```

Note that GeneratePackageOnBuild uses that dreaded `AfterTargets="Build"` causing phasing issues in repos that have multiple projects (e.g. 3 nuget projects, then one down-stream project that validates them all). So when we pack as part of build, we would generally write

```
<PropertyGroup>
  <AfterBuildDependsOn>$(AfterBuildDependsOn);Pack;MyCustomPackageValidation;Etc</AfterBuildDependsOn>
<PropertyGroup>
```

And when migrating projects, this:

```
<Target Name="AfterBuild">
  ...
</Target>
```

Gets renamed and added to AfterBuildDependsOn, e.g.:

```
<PropertyGroup>
  <AfterBuildDependsOn>$(ThisProjectAfterBuild);AfterBuild_</AfterBuildDependsOn>
</PropertyGroup>
<Target Name="AfterBuild_">
   ...
</Target>
```

**Comment by ericsampson** [comment]
> My initial reaction to that plan is utter revulsion, followed by agreeing that we should do it since it's a bunch of gymnastics we can do to avoid user pain. I think that's the plan now.

@rainersigwald @nguerrera or @KirillOsenkov, is this something that could be considered for .NET 6? Given the number of people that are hopefully going to be migrating off Framework onto .NET 6 and potentially run into this little footgun, it would be great to avoid their pain :) Cheers!

**Comment by nguerrera** [comment]
I support this, so thumbs up, but I don't work on .NET anymore so I can't speak for the timeline or anything.

**Comment by trivalik** [comment]
In case this is not fixed I would suggest to give a warning, if easy possible of course.

**Comment by chylex** [comment]
I just ran into this when I happened to name my target `AfterBuild`, and then spent too long trying to figure out why it wasn't running...

I see some people talking about `Overriding target "BeforeBuild" in project` and such appearing in logs, but I've never seen this message in any of my logs. If it appeared, I probably would have figured this out sooner. All I'm seeing is:

### dotnet build
```
Microsoft (R) Build Engine version 16.11.2+f32259642 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

Determining projects to restore...
Restored /home/... (in 296 ms).
TestProject -> /home/...

Build succeeded.
0 Warning(s)
0 Error(s)

Time Elapsed 00:00:01.84 
```

### build log file
```
Build started 12/28/2021 10:36:32 AM.
Logging verbosity is set to: Normal.     1>Project "/home/..." on node 1 (Build target(s)).
     1>ValidateSolutionConfiguration:
         Building solution configuration "Debug|x64".
     1>Project "/home/..." (2) on node 1 (default targets).
     2>GenerateTargetFrameworkMonikerAttribute:
       Skipping target "GenerateTargetFrameworkMonikerAttribute" because all output files are up-to-date with respect to the input files.
       CoreGenerateAssemblyInfo:
       Skipping target "CoreGenerateAssemblyInfo" because all output files are up-to-date with respect to the input files.
       CoreCompile:
         /usr/share/dotnet/dotnet exec (extremely long line follows)
         CompilerServer: server - server processed compilation - f9d8998d-ca95-4d31-bb2c-5b2bc4bf638f
       _CopyOutOfDateSourceItemsToOutputDirectory:
         Copying file from "/home/...".
       GenerateBuildRuntimeConfigurationFiles:
       Skipping target "GenerateBuildRuntimeConfigurationFiles" because all output files are up-to-date with respect to the input files.
       CopyFilesToOutputDirectory:
         Copying file from "/home/...".
         TestProject -> /home/...
         Copying file from "/home/...".
     2>Done Building Project "/home/..." (default targets).
     1>Done Building Project "/home/..." (Build target(s)).

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.19
```

The target just doesn't appear at all in the build logs, with no warning that it was skipped or overridden. Where was I supposed to see the warning?

**Comment by Nirmal4G** [comment]
Moving them to `Common.props` at-least maintains the working state of the upgraded project! It also won't be much of a compat issue since the `Common.targets` already imports `Common.props`!
